### PR TITLE
⚡️ Add chat streaming to OpenAI responses

### PR DESCRIPTION
**Pull request by sabaimran**
# Incoming
- Add chat streaming to the OpenAI API request. Implement using a callback function which manages a queue where new tokens can be placed as they come on. As the thread is read from, tokens are removed.
  - When the final token has been processed, add the `compiled_references` to the queue to be rendered by the `chat` client
  - When the thread has been closed, save the accumulated conversation log in the user's history using a `partial func`
  - Incrementally decode tokens on the front end and add them as they appear from the streamed response
- This significantly reduces perceived latency and will hopefully eliminate incident rate for timeouts

closes https://github.com/khoj-ai/khoj/issues/257

https://github.com/khoj-ai/khoj/assets/65192171/8a67ad26-0b0b-44a9-97a4-4483d7a9f642

### CODE REVIEW

File: src/interface/obsidian/src/chat_modal.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { App, Modal, request, Setting } from 'obsidian';
 import { KhojSetting } from 'src/settings';
-
+import fetch from "node-fetch";
```
Comment (sabaimran) [inline src/interface/obsidian/src/chat_modal.ts:3]: This should work for locally-hosted Khoj + Obsidian clients. @debanjum , I'll need to enlist your help for Emacs support 😁

File: src/interface/obsidian/src/chat_modal.ts [hunk 2]
```diff
@@ -121,10 +156,30 @@ export class KhojChatModal extends Modal {
         // Get chat response from Khoj backend
         let encodedQuery = encodeURIComponent(query);
         let chatUrl = `${this.setting.khojUrl}/api/chat?q=${encodedQuery}&client=obsidian`;
-        let response = await request(chatUrl);
-        let data = JSON.parse(response);
 
-        // Render Khoj response as chat message
-        this.renderMessage(data.response, "khoj");
+        let response = await fetch(chatUrl, {
```
Comment (debanjum) [inline src/interface/obsidian/src/chat_modal.ts:160]: We should add a thinking status message in obsidian as well

**Review by debanjum — COMMENTED** [review]
Tested this out locally. Chat with streaming feels so much better 🌈! Makes the chat feel a lot more snappy and reduces all those OpenAI API call failures

File: src/interface/obsidian/src/chat_modal.ts (hunk 1 above)
Comment (debanjum) [reply src/interface/obsidian/src/chat_modal.ts:3]: Let me try add streaming to `khoj.el` for Emacs separately?

**Review by debanjum — APPROVED** [review]

**Pull request merged**